lib/zkProving.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  MembershipProver,
+  Tree,
+  Poseidon,
+  defaultPubkeyMembershipPConfig,
+} from "@personaelabs/spartan-ecdsa";
+import { loadSigmojis, updateSigmoji } from "@/lib/localStorage";
+import { cardPubKeys } from "./cardPubKeys";
+import { importPublic, ecrecover } from "@ethereumjs/util";
+import { sha256 } from "js-sha256";
+import { serializeSigmojiZKP } from "@/lib/types";
+import { parseDERSignature, findV } from "@/lib/signatureUtils";
+
+export interface ProverWasm {
+  poseidon: Poseidon;
+  prover: MembershipProver;
+}
+
+export async function initWasm(): Promise<ProverWasm> {
+  const poseidon = new Poseidon();
+  await poseidon.initWasm();
+
+  const prover = new MembershipProver({
+    ...defaultPubkeyMembershipPConfig,
+    enableProfiler: true,
+  });
+
+  await prover.initWasm();
+
+  return { poseidon, prover };
+}
+
+export const makeProofs = async (proverWasm: ProverWasm) => {
+  const treeDepth = 20;
+  const pubKeyTree = new Tree(treeDepth, proverWasm.poseidon);
+
+  for (const entry of Object.entries(cardPubKeys)) {
+    const secondaryPublicKeyRawUint8Array = new Uint8Array(
+      (entry[1].secondaryPublicKeyRaw.match(/.{1,2}/g) || []).map((byte) =>
+        parseInt(byte, 16)
+      )
+    );
+    const pubKey = importPublic(secondaryPublicKeyRawUint8Array);
+    const pubKeyBuffer = Buffer.from(pubKey);
+    pubKeyTree.insert(proverWasm.poseidon.hashPubKey(pubKeyBuffer));
+  }
+
+  console.log(pubKeyTree);
+
+  // load in localStorage
+  const sigmojis = await loadSigmojis();
+
+  // generate proofs for all the sigmojis
+  for (const sigmoji of sigmojis) {
+    if (sigmoji.ZKP === "") {
+      // set up signature
+      let { r, s } = parseDERSignature(sigmoji.PCD.proof.cleanedSignature);
+
+      // remove 00 at start of r, s if present
+      if (r.length == 66) {
+        r = r.slice(2);
+      }
+      if (s.length == 66) {
+        s = s.slice(2);
+      }
+
+      // setup pubkey + merkle proof
+      const secondaryPublicKeyRawUint8Array = new Uint8Array(
+        (sigmoji.PCD.claim.pubkeyHex.match(/.{1,2}/g) || []).map((byte) =>
+          parseInt(byte, 16)
+        )
+      );
+      const pubKey = importPublic(secondaryPublicKeyRawUint8Array);
+      const pubKeyBuffer = Buffer.from(pubKey);
+      const pubKeyHash = proverWasm.poseidon.hashPubKey(pubKeyBuffer);
+      const index = pubKeyTree.indexOf(pubKeyHash);
+      const merkleProof = pubKeyTree.createProof(index);
+
+      // set up msgHashArray
+      const rndBuf = Buffer.from(sigmoji.PCD.proof.signedDigest, "hex");
```

Comment:
Seems like this matches the `verify` code in the PCD definition. Might want to refactor this to utils to make it more clear

Reply:
Fair point; I'll leave it here for now since it doesn't get used again.